Documentation/proposals/202405-agent-daemonset.md

```diff
@@ -0,0 +1,132 @@
+
+
+## 6. How
+
+### 6.1. CRD:
+
+Currently, we already have a Prometheus Agent CRD that supports StatefulSet deployment. We’ll add new field(s) in this CRD to enable DaemonSet deployment.
+
+The reason for enhancing existing CRD (instead of introducing a new CRD) is it would take less time to finish the MVP. We’ll let users experiment with the MVP, and in case users report a separate CRD is needed, we’ll separate the logic of DaemonSet deployment into a new CRD later.
+
+The current [Prometheus Agent CRD](https://prometheus-operator.dev/docs/operator/api/#monitoring.coreos.com/v1alpha1.PrometheusAgent) already has sufficient fields for the DaemonSet deployment. The DaemonSet deployment can use all the existing fields in the CRD except the ones related to:
+* Selectors for service, probe, ScrapeConfig
+* Replica
+* Shard
+* Storage
+We can add a new bool field `daemonset`, which is defaulted to `false`. If the DaemonSet deployment is activated (`daemonset: true`), all the fields except the unrelated ones listed above will be used for the DaemonSet mode.
+
+### 6.2. Node detecting:
+
+As pointed out in [Danny from GMP’s talk](https://www.youtube.com/watch?v=yk2aaAyxgKw), to make Prometheus Agent DaemonSet know which node it’s on, we can use [Kubernetes’ downward API](https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/). In `prometheus` or `config-reloader` container, we can mount the node name as an environment variable like this:
+containers:
+```
+ - name: prometheus
+   image: prom/prometheus
+   env:
+   - name: NODE_NAME
+     valueFrom:
+     	fieldRef:
+     		apiVersion: v1
+     		fieldPath: spec.nodeName
+```
+
+### 6.3. Targets filtering for pods (PodMonitor support):
+
+To filter targets, Danny’s talk has pointed out two options.
+The first option is to use relabel config. For example, to filter pods we could generate:
+```
+relabel_configs:
+ - source_labels: [__meta_kubernetes_pod_node_name]
+   separator: ;
+   regex: $NODE_NAME
+   replacement: $1
+   action: keep
+```
+This could be considered as “last-mile filtering”. In other words, we watch for all the changes, and then filter the changes that match.
+
+The second option is to use field selector. For example, to select pods on a specific node, we can do:
+```
+kubernetes_sd_configs:
+   - role: pod
+ 	selectors:
+ 	- role: pod
+   	  field: spec.nodeName=$NODE_NAME
+```
+This option performs better, because we filter targets right at discovery time, and also because Kubernetes API server watch cache indexes pods by node name (as we can see in [Kubernetes codebase](https://github.com/kubernetes/kubernetes/blob/v1.30.0-rc.0/pkg/registry/core/pod/storage/storage.go#L91)). So we’ll go with this option.
```

Comment:
:+1:

Reply:
Could we mention only the option that we choose to move forward here? The other approach could be mentioned in an `Alternatives` section with an explanation of why we decided to ignore it.

Reply:
If we have alternatives for the general approach (covering the whole How section), I'll put them in a separate `Alternatives` section.

But here, it's just an alternative for a part of How. Do you think we need a whole new section for it? I think keeping them at one place/section makes it easier for readers to compare, while keeping them at two sections make readers scroll back and forth to see which part of How this Alternative refers to => harder to compare 🤔

Reply:
I'd be ok briefly mentioning that using relabel configs would be less efficient.

Reply:
> I'd be ok briefly mentioning that using relabel configs would be less efficient.

You mean we should have a separate Alternatives section or compare them together in one section?

Reply:
No need for a separate section. Something along those lines for posterity:

```
### 6.3. Targets filtering for pods (PodMonitor support):

We will use field selectors to filter the pods running on the node. For example, to select pods on a specific node, we can do:

...

We've also evaluated using relabeling configuration but it was deemed less efficient since the agent would retrieve more data from the API server.
```